cpp/src/arrow/flight/client.cc

```diff
@@ -242,140 +244,239 @@ class GrpcClientAuthReader : public ClientAuthReader {
+// To do this, we wrap the gRPC reader in a peekable iterator.
+// The Flight reader can then peek at the message to determine whether
+// it has application metadata or not, and pass the message to
+// RecordBatchStreamReader as appropriate.
+
+// TODO: this deserves some tests of its own with a mock Reader
+template <typename Reader>
+class PeekableFlightDataReader {
  public:
-  GrpcStreamReader();
+  explicit PeekableFlightDataReader(std::shared_ptr<Reader> stream)
+      : stream_(stream), peek_(), finished_(false), valid_(false) {}
 
-  static Status Open(std::unique_ptr<ClientRpc> rpc,
-                     std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream,
-                     std::unique_ptr<GrpcStreamReader>* out);
-  std::shared_ptr<Schema> schema() const override;
-  Status Next(FlightStreamChunk* out) override;
-  void Cancel() override;
+  void Peek(internal::FlightData** out) {
+    *out = nullptr;
+    if (finished_) {
+      return;
+    }
+    if (EnsurePeek()) {
+      *out = &peek_;
+    }
+  }
+
+  void Next(internal::FlightData** out) {
+    *out = nullptr;
+    if (finished_) {
+      return;
+    }
+    if (EnsurePeek()) {
+      *out = &peek_;
+    }
+    valid_ = false;
+  }
 
  private:
-  friend class GrpcIpcMessageReader;
-  std::unique_ptr<ipc::RecordBatchReader> batch_reader_;
-  std::shared_ptr<Buffer> last_app_metadata_;
-  std::shared_ptr<ClientRpc> rpc_;
+  bool EnsurePeek() {
+    if (finished_ || valid_) {
+      return valid_;
+    }
+
+    if (!internal::ReadPayload(&*stream_, &peek_)) {
+      finished_ = true;
+      valid_ = false;
+    } else {
+      valid_ = true;
+    }
+    return valid_;
+  }
+
+  std::shared_ptr<Reader> stream_;
+  internal::FlightData peek_;
+  bool finished_;
+  bool valid_;
 };
 
+// An ipc::MessageReader that adapts any readable gRPC stream
+// returning FlightData.
+template <typename Reader>
 class GrpcIpcMessageReader : public ipc::MessageReader {
  public:
-  GrpcIpcMessageReader(GrpcStreamReader* reader, std::shared_ptr<ClientRpc> rpc,
-                       std::unique_ptr<grpc::ClientReader<pb::FlightData>> stream)
-      : flight_reader_(reader),
-        rpc_(rpc),
+  GrpcIpcMessageReader(std::shared_ptr<ClientRpc> rpc, std::shared_ptr<Reader> stream,
+                       std::shared_ptr<PeekableFlightDataReader<Reader>> peekable_reader)
+      : rpc_(rpc),
         stream_(std::move(stream)),
+        peekable_reader_(peekable_reader),
         stream_finished_(false) {}
 
   Status ReadNextMessage(std::unique_ptr<ipc::Message>* out) override {
+    *out = nullptr;
     if (stream_finished_) {
-      *out = nullptr;
-      flight_reader_->last_app_metadata_ = nullptr;
       return Status::OK();
     }
-    internal::FlightData data;
-    if (!internal::ReadPayload(stream_.get(), &data)) {
-      // Stream is completed
+    internal::FlightData* data;
+    peekable_reader_->Next(&data);
```

Comment:
Does this need to be able to return an error?

Reply:
No, it can only signify that there was/wasn't a message read because gRPC only gives you a boolean status flag - you have to call `Finish()` on the stream to get the server error (if any). I've refactored all the existing logic around that (and the read mutex) into a helper class to manage those details.